File: src/matchbox/web/routes/offers.py
"""JSON API for the offer module (prefix /api/offers).

Exposes CRUD for offers and a truthful salary benchmark against the user's own
job-pool data. See offers/repo.py and offers/benchmark.py.
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel

from matchbox.offers import repo
from matchbox.offers.benchmark import benchmark
from matchbox.web.deps import ConnDep

router = APIRouter(prefix="/api/offers")


class CreateOfferBody(BaseModel):
    applicationId: int
    base: float | None = None
    bonus: float | None = None
    equity: str | None = None
    currency: str | None = None
    location: str | None = None
    receivedAt: str | None = None
    notes: str | None = None


class SetStatusBody(BaseModel):
    status: str


@router.get("")
def list_offers(
    conn: ConnDep,
    applicationId: int | None = Query(default=None),
) -> list[dict[str, Any]]:
    """All offers, or filtered by ?applicationId=N."""
    if applicationId is not None:
        return repo.list_for_app(conn, applicationId)
    return repo.list_all(conn)


@router.post("")
def create_offer(body: CreateOfferBody, conn: ConnDep) -> dict[str, Any]:
    """Record a new offer."""
    oid = repo.create(
        conn,
        body.applicationId,
        base=body.base,
        bonus=body.bonus,
        equity=body.equity,
        currency=body.currency,
        location=body.location,
        received_at=body.receivedAt,
        notes=body.notes,
    )
    created = repo.get(conn, oid)
    assert created is not None  # just inserted
    return created


@router.post("/{offer_id}/status")
def set_offer_status(offer_id: int, body: SetStatusBody, conn: ConnDep) -> dict[str, Any]:
    """Transition an offer to a new status."""
    try:
        offer = repo.set_status(conn, offer_id, body.status)
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    if offer is None:
        raise HTTPException(status_code=404, detail="no such offer")
    return offer


@router.get("/benchmark")
def get_benchmark(
    conn: ConnDep,
    base: float = Query(...),
    roleFamily: str | None = Query(default=None),
    currency: str | None = Query(default=None),
) -> dict[str, Any]:
    """Benchmark a base salary against the user's own discovered-job salary data."""
    return benchmark(conn, base=base, role_family=roleFamily, currency=currency)

File: frontend/src/api/offers.ts
/* Offers API client — talks to the FastAPI backend for the offer ledger and the
   honest own-pool salary benchmark. Negotiation here is status-tracking only:
   the counter-offer draft is written by the user in Claude Code (and voice-
   checked there), never generated in the browser. Every call try/catches with a
   safe fallback so the screen always renders. */

export type OfferStatus = "received" | "negotiating" | "accepted" | "declined";

export interface Offer {
  id: number;
  applicationId: number;
  base: number | null;
  bonus: number | null;
  totalComp: number | null;
  equity: string | null;
  currency: string | null;
  location: string | null;
  status: string;
  receivedAt: string | null;
  notes: string | null;
  createdAt: string;
}

export interface CreateOfferBody {
  applicationId: number;
  base?: number | null;
  bonus?: number | null;
  equity?: string | null;
  currency?: string | null;
  location?: string | null;
  receivedAt?: string | null;
  notes?: string | null;
}

/** Honest benchmark: percentile of `base` against the user's own offer pool.
    `basis` is a verbatim, human-readable description of what was compared; show
    it as-is. When confidence is "none"/sampleSize 0, there is no real percentile. */
export interface Benchmark {
  percentile: number | null;
  sampleSize: number;
  median: number | null;
  range: { low: number; high: number } | null;
  basis: string;
  currency: string | null;
  confidence: "none" | "low" | "medium";
}

const JSON_HEADERS = { "Content-Type": "application/json" };

async function getJSON<T>(url: string, fallback: T): Promise<T> {
  try {
    const r = await fetch(url);
    return r.ok ? ((await r.json()) as T) : fallback;
  } catch {
    return fallback;
  }
}

async function postJSON<T>(url: string, body: unknown, fallback: T): Promise<T> {
  try {
    const r = await fetch(url, {
      method: "POST",
      headers: JSON_HEADERS,
      body: JSON.stringify(body),
    });
    return r.ok ? ((await r.json()) as T) : fallback;
  } catch {
    return fallback;
  }
}

export const listOffers = (applicationId?: number) =>
  getJSON<Offer[]>(
    applicationId === undefined ? "/api/offers" : `/api/offers?applicationId=${applicationId}`,
    [],
  );

export const createOffer = (body: CreateOfferBody) =>
  postJSON<Offer | null>("/api/offers", body, null);

export const setOfferStatus = (id: number, status: OfferStatus) =>
  postJSON<Offer | null>(`/api/offers/${id}/status`, { status }, null);

export function getBenchmark(
  base: number,
  roleFamily?: string,
  currency?: string,
): Promise<Benchmark> {
  const params = new URLSearchParams({ base: String(base) });
  if (roleFamily) params.set("roleFamily", roleFamily);
  if (currency) params.set("currency", currency);
  return getJSON<Benchmark>(`/api/offers/benchmark?${params.toString()}`, {
    percentile: null,
    sampleSize: 0,
    median: null,
    range: null,
    basis: "No matching salary data in your own pool yet.",
    currency: currency ?? null,
    confidence: "none",
  });
}
